File: src/main/kotlin/net/wavem/uvc/rms/gateway/location/response/LocationResponseHandler.kt
package net.wavem.uvc.rms.gateway.location.response

import com.google.gson.Gson
import com.google.gson.JsonObject
import net.wavem.uvc.mqtt.application.MQTTService
import net.wavem.uvc.mqtt.domain.MQTTConnectionType
import net.wavem.uvc.rms.common.domain.header.Header
import net.wavem.uvc.rms.common.types.area.AreaClsfType
import net.wavem.uvc.rms.common.types.header.RobotType
import net.wavem.uvc.rms.common.types.job.JobGroupType
import net.wavem.uvc.rms.common.types.job.JobKindType
import net.wavem.uvc.rms.common.types.job.TaskStatusType
import net.wavem.uvc.rms.gateway.location.domain.Location
import net.wavem.uvc.rms.gateway.location.domain.LocationProperties
import net.wavem.uvc.rms.gateway.location.domain.job.LocationJobInfo
import net.wavem.uvc.rms.gateway.location.domain.last_info.LocationLastInfo
import net.wavem.uvc.rms.gateway.location.domain.position.LocationPosition
import net.wavem.uvc.rms.gateway.location.domain.task.LocationTaskInfo
import net.wavem.uvc.rms.gateway.location.dto.LocationPositionDTO
import net.wavem.uvc.ros.application.topic.Subscription
import net.wavem.uvc.ros.domain.sensor_msgs.NavSatFix
import net.wavem.uvc.ros.domain.sensor_msgs.BatteryState
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

@Component
class LocationResponseHandler(
    private val locationProperties : LocationProperties,
    private val mqttService : MQTTService<String>,
    private val gson : Gson
) {

    private val logger : Logger = LoggerFactory.getLogger(this.javaClass)
    private val rclGPSSubscription : Subscription<NavSatFix> = Subscription()
    private val rclBatteryStateSubscription : Subscription<BatteryState> = Subscription()

    init {
        val rclGPSSubscriptionTopic : String = "/ublox/fix"
        this.rclGPSSubscription.registerSubscription(rclGPSSubscriptionTopic, NavSatFix::class)
        logger.info("RCL {$rclGPSSubscriptionTopic} subscription registered")

        val rclBatteryStateSubscriptionTopic : String = "/battery_state"
        this.rclBatteryStateSubscription.registerSubscription(rclBatteryStateSubscriptionTopic, BatteryState::class)
        logger.info("RCL {$rclBatteryStateSubscriptionTopic} subscription registered")
    }

    private fun buildHeader() : Header {
        return Header(
            robotCorpId = "roc0000001",
            workCorpId = "wco0000001",
            workSiteId = "wst0000001",
            robotId = "rbt0000001",
            robotType = RobotType.AMR.type
        )
    }

    private fun buildJobInfo() : LocationJobInfo {
        val taskInfo : LocationTaskInfo = LocationTaskInfo(
            jobGroup = JobGroupType.SUPPLY.type,
            jobKind = JobKindType.MOVE.type,
            taskStatus = TaskStatusType.ASSIGNED.type
        )

        var jobPlanId : String = ""
        val jobPlanIdFromYML : String? = System.getProperty("jobInfo.jobPlanId")
        if (jobPlanIdFromYML != null) {
            jobPlanId = jobPlanIdFromYML
        }

        var jobGroupId : String = ""
        val jobGroupIdFromYML : String? = System.getProperty("jobInfo.jobGroupId")
        if (jobGroupIdFromYML != null) {
            jobGroupId = jobGroupIdFromYML
        }

        var jobOrderId : String = ""
        val jobOrderIdFromYML : String? = System.getProperty("jobInfo.jobOrderId")
        if (jobOrderIdFromYML != null) {
            jobOrderId = jobOrderIdFromYML
        }

        return LocationJobInfo(
            jobPlanId = jobPlanId,
            jobGroupId = jobGroupId,
            jobOrderId = jobOrderId,
            taskInfo = taskInfo
        )
    }

    private fun buildLastInfo() : LocationLastInfo {
        val locationPositionDTO : LocationPositionDTO = LocationPositionDTO()

        this.rclGPSSubscription.getDataObservable().subscribe {
            val navSatFix : NavSatFix = NavSatFix.read(it)
            locationPositionDTO.xpos = navSatFix.latitude
            locationPositionDTO.ypos = navSatFix.longitude
            locationPositionDTO.heading = 45.0
        }

        this.rclBatteryStateSubscription.getDataObservable().subscribe {
            val batteryState : BatteryState = BatteryState.read(it)
        }

        val locationPosition : LocationPosition = locationPositionDTO.build()

        return LocationLastInfo(
            location = locationPosition,
            areaClsf = AreaClsfType.INDOOR.type,
            floor = "1F",
            batteryLevel = 50,
            velocity = 1.5,
            totalDist = 300
        )
    }

    fun handle() {
        val location : Location = Location(
            header = this.buildHeader(),
            jobInfo = this.buildJobInfo(),
            lastInfo = this.buildLastInfo()
        )

        val locationJSON : JsonObject = gson.toJsonTree(location).asJsonObject

        mqttService.bridge(MQTTConnectionType.TO_RMS, locationProperties.topic, locationJSON.toString())
    }
}

File: src/rms/response/location/handler/location_response_handler.py
import rclpy
import json

from rclpy.node import Node
from rclpy.subscription import Subscription
from rclpy.qos import qos_profile_sensor_data
from rclpy.qos import qos_profile_system_default
from rclpy.callback_groups import MutuallyExclusiveCallbackGroup

from sensor_msgs.msg import NavSatFix

from mqtt import broker
from rms.common.application.uuid_service import UUIDService
from rms.common.application.time_service import TimeService
from rms.common.domain.type.robot.robot_type import RobotType
from rms.common.domain.type.job.job_group_type import JobGroupType
from rms.common.domain.type.job.job_kind_type import JobKindType
from rms.common.domain.type.job.task_status_type import TaskStatusType
from rms.common.domain.type.area.area_clsf_type import AreaCLSFType

from rms.common.domain.header import Header
from rms.response.location.domain.location import Location
from rms.response.location.domain.jobInfo.job_info import JobInfo
from rms.response.location.domain.jobInfo.taskInfo.task_info import TaskInfo
from rms.response.location.domain.lastInfo.last_info import LastInfo
from rms.response.location.domain.lastInfo.location.last_info_location import LastInfoLocation


class LocationResponseHandler():
    rclpy_gps_subscription_topic: str = '/ublox/fix'
    mqtt_location_publisher_topic: str = 'hubilon/atcplus/ros/location'
    
    
    def __init__(self, rclpy_node: Node, mqtt_broker: broker.mqtt_broker) -> None:
        self.rclpy_node = rclpy_node
        
        self.gps_subscription_cb_group: MutuallyExclusiveCallbackGroup = MutuallyExclusiveCallbackGroup()
        self.gps_subscription: Subscription = self.rclpy_node.create_subscription(
            msg_type = NavSatFix,
            topic = self.rclpy_gps_subscription_topic,
            callback = self.rclpy_gps_subscription_cb,
            qos_profile = qos_profile_sensor_data,
            callback_group = self.gps_subscription_cb_group
        )
        
        self.mqtt_broker: broker.mqtt_broker = mqtt_broker
        self.uuid_service: UUIDService = UUIDService()
        self.time_service: TimeService = TimeService()
        
        self.location_xpos: float = 0.0
        self.location_ypos: float = 0.0
        self.heading: float = 45.0
        
        
    def rclpy_gps_subscription_cb(self, gps_cb: NavSatFix) -> None:
        self.rclpy_node.get_logger().info(
            'Location GPS cb\n\tlat : "%f"\n\tlon : "%f"\n\talt : "%f"' % (
                gps_cb.latitude,
                gps_cb.longitude,
                gps_cb.altitude
            )
        )
        self.location_xpos = gps_cb.latitude
        self.location_ypos = gps_cb.longitude
        self.heading = gps_cb.altitude
    
    
    def build_location(self) -> Location:
        header: Header = self.__build_header__()
        job_info: JobInfo = self.__build_job_info__()
        last_info: LastInfo = self.__build_last_info__()
        
        location: Location = Location(
            header = header.__dict__,
            jobInfo = job_info.__dict__,
            lastInfo = last_info.__dict__
        )
        
        return location
        
        
    def __build_header__(self) -> Header:
        header: Header = Header(
            robotCorpId = 'rco0000001',
            workCorpId = 'wco0000001',
            workSiteId = 'wst0000001',
            robotId = 'rbt0000001',
            robotType = str(RobotType.AMR)
        )

        return header
    

    def __build_job_info__(self) -> JobInfo:
        job_plan_id: str = self.uuid_service.generate_uuid()
        job_group_id: str = self.uuid_service.generate_uuid()
        job_order_id: str = self.uuid_service.generate_uuid()
        
        task_info: TaskInfo = TaskInfo(
            jobGroup = str(JobGroupType.SUPPLY),
            jobKind = str(JobKindType.MOVE),
            taskStatus = str(TaskStatusType.ASSIGNED)
        )
        
        job_info: JobInfo = JobInfo(
            jobPlanId = job_plan_id,
            jobGroupId = job_group_id,
            jobOrderId = job_order_id,
            taskInfo = task_info.__dict__
        )
        
        return job_info
    
    
    def __build_last_info__(self) -> LastInfo:
        last_info_location: LastInfoLocation = LastInfoLocation(
            xpos = self.location_xpos,
            ypos = self.location_ypos,
            heading = self.heading
        )
        
        last_info: LastInfo = LastInfo(
            location = last_info_location.__dict__,
            areaClsf = str(AreaCLSFType.INDOOR),
            floor = "1F",
            batteryLevel = 50,
            velocity = 1.5,
            totalDist = 300
        )
        
        return last_info
    
    
    def response_to_uvc(self) -> None:
        built_location: Location = self.build_location()
        self.mqtt_broker.client.publish(
            self.mqtt_location_publisher_topic,
            json.dumps(built_location.__dict__)
        )